### ISSUE TITLE
Refatorar: `Client` em `Client` e `AsyncClient`

### ISSUE THREAD

**Issue by cuducos** [FelipeSBarros/crossfire#66]
Mantendo a lógica em `AsyncClient`, e envelopando para uso sem async no `Client`.

**Comment by FelipeSBarros** [comment]
@cuducos a proposta seria fazer o que foi feito em #68 ?

**Comment by cuducos** [comment]
Não. No #68 não temos um cliente síncrono que tem como papel envelopar a lógica de um cliente assíncrono.

Pense que `Client` não deve ter *nenhum* método assíncrono (nada de `async def`). Já o `AsyncClient` não deve ter nenhum método público síncrono (ou seja, nada de `def states` ou `def cities` ou `def occurrences`, mas  `async def states` e `async def cities` e `async def occurrences`).

Pense num caso em que alguém quer usar o pacote assincronamente, por exemplo:


```python
client = AsyncClient()
occurrences = await client.occurrences(‘…’)
```

**Comment by FelipeSBarros** [comment]
@cuducos , se vc não se importa, vou tentar tirar o máximo de perguntas antes de mexer no código. Quero tentar, dentro do possível, fazer essa alteração com o mínimo de commits (pois imagino que, ao fazer assim, me aproximo dessa ideia de modificação "atomica", específica).

Vamos lá ("pensando alto" e revisando o código pois sinto que perdi o "controle dele"):

O uso atual, a partir do #68 seria, para casos simples:

```python
from crossfire import cities, occurrences, states

states()  # lista UFs
cities() # lista municipios
occurrences('2a98a020-3815-45d7-a6f6-6de2119eba8b')  # lista ocorrências
```
Ou seja, de forma geral, entendo que o que foi feito foi a simplificação do uso.
O `corssfire.__init__` passou a "envelopar" o `Client` de `crossfire.clients`, que possui métodos "principais" asíncronos.

Logo, até o momento, o pacote trabalha com a lógica asíncrona, exclusivamente. **Não existe a possibilidade de usar a lógica síncrona**. 

Entendo que este é o seu ponto, em:
> Pense num caso em que alguém quer usar o pacote assincronamente, por exemplo:

Portanto, a proposta desse PR sería:
1. refatorar o `Client`atual em `AsyncClient` (já que atualmente ele já está como asíncrono, e criar uma nova classe `Client` com seus respectivos métodos usando a abordagem síncrona.
    
Neste ponto, tenho uma pergunta: Considerando essa abordagem de transformar a classe `Client` atual em `AsyncClient` não preciso me preocupar com o comentário aseguir, já que a classe em questão apenas apresenta métodos público asíncronos. Confere?
> Já o AsyncClient não deve ter nenhum método público síncrono (ou seja, nada de def states ou def cities ou def occurrences, mas async def states e async def cities e async def occurrences).

Esse minha dúvida reside no fato de `Client.states()`, por exemplo, em sí ser um método que "envelopa" o uso do `self._states`. Sendo este útimo, sim, asíncorno... 

2. criar classe `Client` e respectivos métodos com abordagem síncrona;

**Confere?**

Só fico com uma "pulga atrás de orelha": Por qual motivo alguém usaria o pacote de forma síncrona?

Se, de fato, a ideia é criar uma classe `Client` síncrona, como você sugere que seja feita? Lembro que inicialmente estávamos fazendo como um `generator` (não lembro se desde o ínicio usamos a abordagem asíncrona, acho que não). 

Ou sugere outra abordagem?

Estou deixando passar algum ponto importante?

**Comment by cuducos** [comment]
Me deu a impressão que você está confundindo como _o pacote trabalha_ (por detrás do capô) e como _o pacote pode ser utilizado_.

Não vamos mudar como _o pacote trabalha_ (por detrás do capô). Iremos mudar como _o pacote pode ser utilizado_.

Minha ideia original quando você me pediu sugestão foi algo como:

```python
from crossfire import Client


client = Client()
client.states()
```

Aí você falou que isso poderia ser complexo, pensando em pessoas que nunca trabalharam com classes e instâncias de classes.

O PR #68 não mudou nada de como _o pacote trabalha_ (por detrás do capô), mas ofereceu a quem instala o pacote uma forma simplificada de como _o pacote pode ser utilizado_:

```python
from crossfire states


states()
```

**Pergunta 1**: está claro a diferença entre  como _o pacote trabalha_ (por detrás do capô) e como _o pacote pode ser utilizado_?

**Pergunta 2**: está claro que o #68 não mudou o como _o pacote trabalha_ (por detrás do capô), mas apenas o como _o pacote pode ser utilizado_?

Se sim… da mesma forma que você me trouxe a dúvida sobre pessoas com menos experiência com Python, eu trago, nessa _issue_ aqui, outro “e se“: e se alguém com mais experiência quiser fazer requisições paralelas, por exemplo, pegar todas as ocorrências de todos as UFs? Para isso precisamos “expor” essa forma de como utilizar, permitindo um uso como:


```python
from crossfire import AsyncClient
    

async def main():
    client = AsyncClient()
    states = await client.states()
    tasks = tuple(client.occurrences(state['id']) for state in states)
    data = gather(*tasks)  # isso faz as requisições para cada estado em paralelo
    print(data)
```

Isso, hoje, é quase impossível pois não  expomos formas assíncronas de fazer as consultas por UFs, cidades ou ocorrências.

**Comment by FelipeSBarros** [comment]
> **Pergunta 1**: está claro a diferença entre como _o pacote trabalha_ (por detrás do capô) e como _o pacote pode ser utilizado_?

Perfeito. Está claro, sim. 

> Pergunta 2: está claro que o https://github.com/FelipeSBarros/crossfire/pull/68 não mudou o como o pacote trabalha (por detrás do capô), mas apenas o como o pacote pode ser utilizado?

Também está claro. Por isso disse, na msg anterior:

> O uso atual, a partir do #68 seria, para casos simples:
> 
> ```python
> from crossfire import cities, occurrences, states
> 
> states()  # lista UFs
> cities() # lista municipios
> occurrences('2a98a020-3815-45d7-a6f6-6de2119eba8b')  # lista ocorrências
> ```
> 
> Ou seja, de forma geral, entendo que o que foi feito foi a simplificação do uso.



> e se alguém com mais experiência quiser fazer requisições paralelas, por exemplo, pegar todas as ocorrências de todos as UFs?

Perfeito! Entendi o uso que seria dado! Não tinha pensado nisso. 

1. Voltando aos conceitos "uso" e "fucnionamento", estamos falando em formas de uso, confere? (o trecho de código que vc comprtilhou me faz creer que seja isso.)

> Isso, hoje, é quase impossível pois não expomos formas assíncronas de fazer as consultas por UFs, cidades ou ocorrências.

Acho que entendi. 

2. A ideia é ter uma forma de usar o pacote a partir da qual seria possível acessar as ocorrencias de todos (ou uma lista de estados e/ou cidades)?

> ```python
> from crossfire import AsyncClient
>     
> 
> async def main():
>     client = AsyncClient()
>     states = await client.states()
>     tasks = tuple(client.occurrences(state['id']) for state in states)
>     data = gather(*tasks)  # isso faz as requisições para cada estado em paralelo
>     print(data)
> ```

Entendo que esse seria o código **usado** por alguém para baixar todas as ocorrencias de todos os estados de forma asíncrona, conforme a sua sugestão. E não o código q ser implementado. Confere?

De qualquer forma, entendo que a solução seja a criação de uma nova classe `AsyncClient`, a qual todos os métidos seriam asíncronos.
Olhando o código de `cients.__init__`, vejo que todos os métodos expostos (`states`, `cities` e `occurrences`), são "empacotamentos" que usam `get_event_loop` para acessar resultado das funções não expostas asíncronas (imagino eu por terem `_` como prefixo), essas últimas iniciadas por `async def...`.

Exemplo:
````python
#clients.__init__
...
    async def _states(self, format=None):
        return await self.get(f"{self.URL}/states", format=format)

    def states(self, format=None):
        loop = get_event_loop()
        states, _ = loop.run_until_complete(self._states(format=format))
        return states
```

**Faz sentido isso que estou dizendo ou realmente, não entendi nada?**

Se entendi bem, seria o caso de ter uma nova classe `AsyncClient` que tenha o `states` definido por `awat def states(...)`, para que o código de uso que vc apondou possa ser usado com `await`.

Seria isso?

**Comment by cuducos** [comment]
Sim. Tirando uns _typos_ (`async` ao invés de `awat`, por exemplo) é algo assim mesmo.

**Comment by FelipeSBarros** [comment]
> Sim. Tirando uns _typos_ (`async` ao invés de `awat`, por exemplo) é algo assim mesmo.

Perfeito. editei a msg. 
Vou assistir ao video sugerido ( [API driven devlopment](https://www.youtube.com/watch?v=JQBnopNPQEc) ). Nãose sei de vai me ajudar, mas não sei bem como implementar o que está sendo proposto....
Já volto com mais dúvidas.

**Comment by cuducos** [comment]
> Nãose sei de vai me ajudar, mas não sei bem como implementar o que está sendo proposto....

Resumindo:
* O `Client` atual pode ser renomeado para `AsyncClient`
* Os métodos síncronos “empacotados” (como vc chamou) vão para uma nova classe `Client` que aproveita a lógica da `AsyncClient`

**Comment by FelipeSBarros** [comment]
Ok, vou atacar essa issue pela forma sugerida:

> * O `Client` atual pode ser renomeado para `AsyncClient`

Antes disso, dei uma revisada no código e vi que temos uma classe do `httpx` que se chama `AsyncClient`.
Logo, teriamos que usar outro nome ou usar um alias para o AsyncClient.
Qual abordagem seria melhor?

**Comment by cuducos** [comment]
> Logo, teriamos que usar outro nome ou usar um alias para o AsyncClient.

Ou mudar de `from httpx import AsyncClient` para `import httpx`.

Diria que não faz sentido usar outor nome para a nossa classe se temos _duas_ formas de evitar o conflito de nomes.

Qual das duas, deixo a teu critério : )
